packages/cipherstash-proxy/src/config/database.rs

```diff
@@ -0,0 +1,91 @@
+use std::{fmt::Display, time::Duration};
+
+use rustls_pki_types::ServerName;
+use serde::Deserialize;
+
+use crate::error::{ConfigError, Error};
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct DatabaseConfig {
+    #[serde(default = "DatabaseConfig::default_host")]
+    pub host: String,
+
+    #[serde(default = "DatabaseConfig::default_port")]
+    pub port: u16,
+
+    pub name: String,
+    pub username: String,
+    pub password: String,
+
+    #[serde(default = "DatabaseConfig::default_connection_timeout")]
+    pub connection_timeout: u64,
+
+    #[serde(default)]
+    pub with_tls_verification: bool,
+
+    #[serde(default = "DatabaseConfig::default_config_reload_interval")]
+    pub config_reload_interval: u64,
+
+    #[serde(default = "DatabaseConfig::default_schema_reload_interval")]
+    pub schema_reload_interval: u64,
+}
+
+impl DatabaseConfig {
+    pub fn default_host() -> String {
+        "127.0.0.1".to_string()
+    }
+
+    pub fn default_port() -> u16 {
+        5432
+    }
+
+    // 5 minutes
+    pub fn default_connection_timeout() -> u64 {
+        1000 * 60 * 5
+    }
+
+    pub fn default_config_reload_interval() -> u64 {
+        60
+    }
+
+    pub fn default_schema_reload_interval() -> u64 {
+        60
+    }
+
+    pub fn to_socket_address(&self) -> String {
+        format!("{}:{}", self.host, self.port)
+    }
+
+    pub fn to_connection_string(&self) -> String {
+        format!(
+            "postgres://{}:{}@{}:{}/{}",
+            self.username, self.password, self.host, self.port, self.name
+        )
+    }
+
+    pub fn connection_timeout(&self) -> Duration {
+        Duration::from_millis(self.connection_timeout)
+    }
+
+    pub fn server_name(&self) -> Result<ServerName, Error> {
+        let name = ServerName::try_from(self.host.as_str()).map_err(|_| {
+            ConfigError::InvalidServerName {
+                name: self.host.to_owned(),
+            }
+        })?;
+        Ok(name)
+    }
+}
+
+///
+/// Password is NEVER EVER displayed
+///
+impl Display for DatabaseConfig {
```

Comment:
I recommend the use of [opaque-debug](https://crates.io/crates/opaque-debug).
Or could be a good opportunity to use [vitaminc-protected](https://crates.io/crates/vitaminc-protected).

Reply:
Pushed change using vitaminc for database password.